### DISCUSSION TITLE
Inconsistent truncation of timing points

### DISCUSSION THREAD

**Discussion by JPK314** [ppy/osu#19166] [Q&A]
### Type

Game behaviour

### Bug description

I am developing a script that converts v128 .osu files to v14 without changing how the beatmap plays. One thing my users have reported to me is that slider ticks are sometimes messed up on objects. We traced and examined the issue, it looks like the following is happening:

Create two timing points, A and B, where A is at least one millisecond before B, and where the time of B is not an integer. Place a slider H at the time of B, truncated to an integer.

When H is loaded, three relevant factors are calculated - the current bpm, the current sv multiplier, and the slider tick spacing.

The current bpm is calculated using the most recent uninherited timing point whose truncated time is on or before the time of H.

The current sv is calculated using the most recent inherited timing point (or 1 if none exists) whose truncated time is on or before the time of H.

The slider tick spacing is calculated using the current bpm (as calculated above) and the sv of the most recent inherited timing point (or 1 if none exists) whose *NON-TRUNCATED* time is on or before the time of H. 

Specifically, this means the slider tick spacing is no longer tied to the slider velocity H uses in the case of B being inherited, which is a problem.

I believe this issue arose fairly recently - perhaps as a quick patch for this issue? https://github.com/ppy/osu/issues/17258

The obvious solution is to fix the tick spacing calculation to use the sv of the most recent truncated timing point that is on or before H, as that makes everything consistent.

However, that's boring! It would be pretty cool to be able to control tick spacing and slider velocity independently, and this bug is very close to being able to do that. If the slider tick spacing used the bpm from the most recent uninherited timing point whose non-truncated time is before or on the time of H, this would also make everything consistent and would allow for the independent control described above. For people like me who do not map for ranked or universal compatibility, having bpm changes for mapping purposes instead of to match the song is pretty normal, and I could have a bit of fun with this. You can feel free to ignore this part, but I felt like I had to suggest it. Thanks for your time.

### Screenshots or videos

![unknown](https://user-images.githubusercontent.com/25115269/179369545-ba23d32e-9da4-4c83-97f6-576696ea1181.png)

### Version

b20220424

### Logs

Not applicable

**Comment by peppy** [comment]
Most likely a duplicate of https://github.com/ppy/osu/issues/17258.

> I am developing a script that converts v128 .osu files to v14 without changing how the beatmap plays. One thing my users have reported to me is that slider ticks are sometimes messed up on objects.

If your goal is to contribute to the ecosystem, then making this work inside of lazer is optimal, since that is what we are aiming for.

**Reply by JPK314** [reply]
I mention that issue in my post. This isn't a duplicate of that, this is an issue entirely inside stable. I only mentioned my script to explain how I found the issue - I have no intentions of putting it in lazer as I hate doing UI work. Plus, converting v128 to v14 has already been discussed in https://github.com/ppy/osu/discussions/18066

**Reply by peppy** [reply]
Fair enough. I'll track this alongside #18066 because it's going to be about the same effort to get deep enough to fix one or both (assuming they require different fixes).
